src/common/stats/StatsManager.cpp

```diff
@@ -309,6 +321,214 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
+    } else if (parts[2] == "600") {
+        range = TimeRange::TEN_MINUTES;
+    } else if (parts[2] == "3600") {
+        range = TimeRange::ONE_HOUR;
+    } else {
+        // Unsupported time range
+        LOG(ERROR) << "Unsupported time range \"" << parts[2] << "\"";
+        return Status::Error(folly::stringPrintf("Unsupported time range \"%s\"",
+                                                 parts[2].c_str()));
+    }
+
+    StatsMethod method;
+    folly::toLowerAscii(parts[1]);
+    if (parts[1] == "sum") {
+        method = StatsMethod::SUM;
+    } else if (parts[1] == "count") {
+        method = StatsMethod::COUNT;
+    } else if (parts[1] == "avg") {
+        method = StatsMethod::AVG;
+    } else if (parts[1] == "rate") {
+        method = StatsMethod::RATE;
+    // } else if (parts[1][0] == 'p') {   // TODO(shylock)
+    } else {
+        LOG(ERROR) << "Unsupported statistic method \"" << parts[1] << "\"";
+        return Status::Error(folly::stringPrintf("Unsupported statistic method \"%s\"",
+                                                 parts[1].c_str()));
+    }
+
+    return StatsManager::ParsedName {
+        name,
+        method,
+        range,
+    };
+}
+
+
+void PrometheusSerializer::annotate(
+        std::ostream& out, const std::string& metricName, const std::string& metricType) const {
+    out << "# HELP " << metricName << " Record all " << metricType << " about nebula" << "\n";
+    out << "# TYPE " << metricName << " " << metricType << "\n";
+}
+
+// Write a double as a string, with proper formatting for infinity and NaN
+void PrometheusSerializer::writeValue(std::ostream& out, double value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        auto saved_flags = out.setf(std::ios::fixed, std::ios::floatfield);
+        out << value;
+        out.setf(saved_flags, std::ios::floatfield);
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const int64_t value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        out << value;
+    }
+}
+
+void PrometheusSerializer::writeValue(std::ostream& out, const std::string& value) const {
+    for (auto c : value) {
+        switch (c) {
+        case '\n':
+            out << '\\' << 'n';
+            break;
+
+        case '\\':
+            out << '\\' << c;
+            break;
+
+        case '"':
+            out << '\\' << c;
+            break;
+
+        default:
+            out << c;
+            break;
+        }
+    }
+}
+
+template<typename T>
+void PrometheusSerializer::writeHead(std::ostream& out, const std::string& family,
+        const std::vector<Label>& labels, const std::string& suffix,
+        const std::string& extraLabelName,
+        const T extraLabelValue) const {
```

Comment:
Template member function is preferably defined in the header file.

Reply:
This will not called outside the source file.

Reply:
If so, Make it private!

Reply:
Yes.